docs/Troubleshooting.md

```diff
@@ -37,7 +33,31 @@ function todos(state = [], action) {
 }
 ```
 
-It needs to be rewritten like this:
+Redux depends on you never mutating the objects it passes to your reducer, or objects nested within them. **Every single time you change state, you must return a new state object.** Even if you don’t use a library like [Immutable](https://facebook.github.io/immutable-js/), you need to completely avoid mutation.
+
+Immutability is what lets [react-redux](https://github.com/gaearon/react-redux) efficiently subscribe to fine-grained updates of your state. It also enables great developer experience features such as time travel with [redux-devtools](http://github.com/gaearon/redux-devtools).
+
+To avoid mutating arrays you can use things like `.slice()`, `.concat()`, and spread (`...state`) to extract the elements you want to retain and combine them with any new elements you may want to add, while omitting elements you don’t want.
+
+To avoid mutating objects you’ll generally want to use a utility like `extend()` from Underscore, `assign()` from Lodash, or [`Object.assign()`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) or a polyfill for it.
+
+`Object.assign()` mutates its first argument, so make sure that you use it correctly. For example, instead of returning something like `Object.assign(state, { completed: true })` from your reducers, return `Object.assign({}, state, { completed: true })`. This way you don’t mutate the previous `state`.
+
+If you’re using Babel 6 you can alternatively use the [proposed object spread syntax](https://github.com/sebmarkbage/ecmascript-rest-spread) by enabling [`babel-plugin-transform-object-rest-spread`](http://babeljs.io/docs/plugins/transform-object-rest-spread/) or a preset that includes it. (At the time of this writing the proposal is stage 2 and therefore the plugin is included directly in [`babel-preset-stage-2`](http://babeljs.io/docs/plugins/preset-stage-2/) and indirectly in lower numbered stage presets.) Example:
```

Comment:
I’d say let’s wait for #1358 to get merged and just link to it rather than explain it in detail here.